// typescript.events/lib/server/dictionary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Dictionary = /** @class */ (function () {
    function Dictionary() {
        this.items = {};
        this.count = 0;
    }
    Dictionary.prototype.Add = function (key, value) {
        this.items[key] = value;
        this.count++;
    };
    Dictionary.prototype.ContainsKey = function (key) {
        return this.items.hasOwnProperty(key);
    };
    Dictionary.prototype.Count = function () {
        return this.count;
    };
    Dictionary.prototype.Item = function (key) {
        return this.items[key];
    };
    Dictionary.prototype.Keys = function () {
        var Keys = [];
        // tslint:disable-next-line:forin
        for (var key in this.items) {
            Keys.push(key);
        }
        return Keys;
    };
    Dictionary.prototype.Remove = function (key) {
        var val = this.items[key];
        delete this.items[key];
        this.count--;
        return val;
    };
    Dictionary.prototype.Values = function () {
        var values = [];
        // tslint:disable-next-line:forin
        for (var key in this.items) {
            values.push(this.items[key]);
        }
        return values;
    };
    return Dictionary;
}());
exports.Dictionary = Dictionary;
